Add CEPH manifests

xarses opened the pull request:
Adds manifests to deploy Ceph parts.

Commit:
import from internal@04f65450b945f8614c8fbeb4554a2b993d1fcbce

Commit:
first attempt to convert to support yum insteal of apt

Commit:
fixed refrence to existing vars; added missing rhel packages; fixed yum repo vars; added required stdlib for suffix

Commit:
added suffix files

noskovao committed:
Fix service cinder restart

Commit:
updated readme; added some hack parts to site.pp; added some cleanup notes to site.pp;

Commit:
moved all yum repos to yum.pp

Commit:
cleaned package requirements up slightly

Commit:
fix suoders !requiretty problem

Commit:
fixed requiretty file

Commit:
move ssh key from file to ssh_authorized_key builtin

Commit:
[PRD-1544] define firewall rules for ceph

added firewall rules
moved deps away from example/site.pp

Commit:
[PRD-1536] rewrote README.rst

refactored troubleshooting examples and steps
refactored parameter defs

Commit:
[PRD-1542] seperate mon and osd list

sperated vars
rewrote vars
converted cinder / glance to use RHEL correctly
fixed deploy to deploy ceph correctly for RHEL
removed disk zap from deploy

Commit:
fix errors in ssh.pp

Commit:
fix deps class

Commit:
convert deps to resource

Commit:
move fw rules out of deps.pp

Commit:
fix cinder service name

Commit:
[PRD-1536] fixed some ts examples

Commit:
cleaned up some puppet errors

noskovao committed:
RadosGW for CEPH

Commit:
fixed race condition between ceph-deploy mon and ceph-deploy osd

Commit:
fix markup in README.rst

angdraug committed:
README converted to markdown and cleaned up

angdraug committed:
README update with Glance checks

Commit:
fixup minor deployment errors

fixed examples and markup in readme
fix cinder config file issues, cleaned wrong return states
fix glance wrong return states
fix deploy.pp timing
fix deploy.pp auth bug

Commit:
fix run order in site.pp

Commit:
disable external repos

mirantis-jenkins commented:
Can one of the admins verify this patch?

xarses commented:
@xdel the ceph-deploy script makes extensive use of sudo, even though its running as root. Also the ceph-deploy scripts require ssh-keys ceph/files/openstack\* that need to be on every controller, storage and compute.

xarses commented:
@xdel yum.pp isn't called currently because the packages are in the repo and is currently not being called.

xdel commented on deployment/puppet/ceph/manifests/deploy.pp:
```diff
@@ -0,0 +1,89 @@
+  class c_pools {
+    define int {
+      exec { "Creating pool ${name}":
+        command => "ceph osd pool create ${name} 128",
```
pg_num and pgp_num should be multiplied by number of computes

xdel replied:
Oh I said computes.. cinders ofc. Please think about 50 pgs/node with assumption we`ll have two or three pools with heavy amounts of data in observable future.

xarses replied:
See http://ceph.com/docs/next/rados/operations/pools/#createpool and http://ceph.com/docs/next/rados/operations/placement-groups/ they imply that they shouldn't be changed aftewards so we will need a better way to calculate what we think we want this value to be. At a minimum, this should be changed to a parameter

xdel replied:
User probably don`t know how to change this parameter properly without such knowledge so at first place there should be an automated variable and we can expose it to the parent in remaining time.

xdel commented on deployment/puppet/ceph/manifests/cinder.pp:
```diff
@@ -0,0 +1,52 @@
+      require => [Package['ceph'], Exec['Copy configs']],
+      returns => 0,
+    }
+    file { '/etc/ceph/ceph.client.volumes.keyring':
```
please put safer permissions on every keyring not created by ceph-deploy

xdel commented on deployment/puppet/ceph/examples/site.pp:
```diff
@@ -0,0 +1,83 @@
+  if $fqdn == $mon_nodes[-1] and !str2bool($::ceph_conf) {
+    class { 'ceph::deploy':
+      auth_supported   => 'cephx',
+      osd_journal_size => '2048',
```
should be configurable via user parameter in the parent manifest

xdel commented on deployment/puppet/ceph/manifests/cinder.pp:
```diff
@@ -0,0 +1,52 @@
+  if str2bool($::cinder_conf) {
+
+    exec {'Copy configs':
+      command => "scp -r ${mon_nodes[-1]}:/etc/ceph/* /etc/ceph/",
```
Do you think this is really needed? Ceph-deploy should take care of this.

xarses replied:
hold over from original script. This wouldn't be needed for nodes with osd or mon roles. I think the use case was for mon/osd not being on the node that cinder and glance where being configured on. Assuming mon's on controllers and osd's on storage nodes; this would be necessary for the compute nodes.

xdel replied:
Hmm.. Ceph-deploy will spread config from ''initial'' node everywhere(or it needs to :) ), so I`m in doubt if such action needed anyway.

xarses replied:
the current logic (which i want to re-factor) is to setup each of the mon's then run ceph-deploy on the last one, so the key's aren't present until mon[-1].

xdel commented on deployment/puppet/ceph/manifests/nova_compute.pp:
```diff
@@ -0,0 +1,26 @@
+      content => template('ceph/secret.erb')
+    }
+    exec { 'Set value':
+      command => 'virsh secret-set-value --secret $(virsh secret-define --file /tmp/secret.xml | egrep -o "[0-9a-fA-F]{8}(-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}") --base64 $(ceph auth get-key client.volumes) && rm /tmp/secret.xml',
```
Holy cow!

xarses replied:
other than the egrep this is from the setup docs http://ceph.com/docs/next/rbd/rbd-openstack/

xdel replied:
I meant putting ceph key in the tmp without mktemp or something alike looks awful. Predefined secret looks very bad too(you have get-or-create for such situation).

xdel commented on deployment/puppet/ceph/manifests/ssh.pp:
```diff
@@ -0,0 +1,40 @@
+  }
+  file { '/etc/ssh/ssh_config':
+    mode    => '0600',
+    content => "Host *\n  StrictHostKeyChecking no\n  UserKnownHostsFile=/dev/null\n",
```
This file should be replaced by our ssh magic soon.

xarses replied:
yes it should =)

plechenko commented:
add to whitelist

mirantis-jenkins commented:
:-1: Test FAILed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel-syntax-check/14/

angdraug commented:
Test failed due to github error, please retest.

plechenko commented:
retest this please

mirantis-jenkins commented:
:-1: Test FAILed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel-syntax-check/19/

Commit:
multiple fixes

fixed broken cinder -> ceph
fixed keyring modes
added hack to ceph-deploy to push keys to other monitors so they dont have
 to run puppet again. This needs to be removed by PRD-1570

xdel commented on deployment/puppet/ceph/manifests/deploy.pp:
```diff
@@ -0,0 +1,89 @@
+        command => "ceph-deploy osd prepare ${devices}",
+        returns => 0,
+        timeout => 0, #TODO: make this something reasonable
+        tries => 60,  #This is necessary because of race for mon creating keys
```
What is this thing?

xarses replied:
ceph-deploy osd prepare is ensuring there is a filesystem on the disk according to the args passed to ceph.conf (above). 

timeout: It has a long timeout because of the format taking forever. A resonable amount of time would be around 300 times the length of $osd_nodes. Right now its 0 to prevent puppet from aborting it.

tries: there is a race condition that this is preventing, after the 'ceph-deploy create mon' the bootstrap and admin keys must exist before the osd commands can be issued. This however requires the mon node to establish a quorum prior which takes 10-15 seconds for me. because 'ceph-deploy create mon' or 'ceph-deploy gather-keys' doesn't block, this seemed like an OK place to block, but it really should be in with 'ceph-deploy gather-keys'

angdraug replied:
Usually when code needs explanations that long, it's worth putting the explanations in the comments near the code, so that no one tries to refactor it away later without fully understanding why it was done this way in the first place.

xdel replied:
May be we cay try to rely on the output of ceph-deploy and fail if nothing comes out in relatively large time interval, instead of doing this?

xarses replied:
ceph-deploy create mon dosen't block for the quorum, we have to wait somewhere. Are you proposing that we monitor the log? In the mean time, I moved this to a new block using ceph-deploy gather-keys instead of 'osd prepare'

xdel commented on deployment/puppet/ceph/manifests/deps.pp:
```diff
@@ -0,0 +1,12 @@
+  $type = 'base'
+){
+  
+  package { ['ceph', 'redhat-lsb-core','ceph-deploy', 'python-pushy']:
```
Let`s install everything including rgw, it may shorten our job during future upgrades.

xarses replied:
rgw support is coming in to a push near you ... soon.

xdel commented on deployment/puppet/ceph/manifests/ssh.pp:
```diff
@@ -0,0 +1,40 @@
+  $ssh_private_key = 'puppet:///modules/ceph/openstack'
+  $ssh_public_key = 'puppet:///modules/ceph/openstack.pub'
+
+  File {
```
What file?

xarses replied:
because its a capital F, I think its setting resource defaults for the remaining "file" declarations, It's a hold over from the import from the internal repo.

mirantis-jenkins commented:
:-1: Test FAILed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel-syntax-check/20/

Commit:
clean from rebase to add radiosgw

xarses commented:
rebased  087c323 and 719303e from internal into this branch to add radiosgw support added 3c7d07d to clean up the rebase

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel-syntax-check/24/

Commit:
multiple fixes

remove tmp file when done in nova_compute
annotate some of the deploy commands in deploy
fixed some of the formating in deploy
added some task marks

xdel commented:
+1

mirantis-jenkins commented:
:+1: Test PASSed.
Refer to this link for build results: http://fuel-jenkins.mirantis.com/job/fuel-syntax-check/26/

bogdando commented on deployment/puppet/ceph/README.md:
```diff
@@ -0,0 +1,307 @@
+2. Copy ``fuel-pm:/etc/puppet/modules/*`` to ``{node}:/etc/puppet/modules``
+3. Copy ``/etc/puppet/modules/ceph/examples/site.pp`` to ``/root/ceph.pp``.
+4. Edit ceph.pp for desired changes to ``$mon_nodes``, ``$osd_nodes``, and ``$osd_disks``.
+5. Run ``puppet apply ceph.pp`` on each node **except** ``$ceph_nodes[-1]``,
```
What does it mean? Except last one, but "then run the same command on that last one". When then?

angdraug replied:
It means that `puppet apply ceph.pp` should be run on `$ceph_nodes[-1` only after `puppet apply ceph.pp` has finished on all other nodes. I'll create a patch to clarify the wording here.
